### PR TITLE
[SYCL]Matching kernel's function types with actual types.

### PR DESCRIPTION
**Pull request by zahiraam**
Using getFullyQualifiedType to calculcate the kernel function's types makes the
types in the kernel different from the actual type and that creates a type
mismatch.
getFullyQualifiedType's use is restricted to make type lookup correct in
templates.
This change, replacing the fully qualified type with the canonical
one, fixes the issue, but will imply that the pretty-printer needs to be
updated to print fully qualified names.

### CODE REVIEW

File: clang/test/CodeGenSYCL/kernelname-enum.cpp [hunk 1]
```diff
@@ -208,3 +208,4 @@ int main() {
 // CHECK: template <> struct KernelInfo<::T1<::T2<static_cast<::type_argument_template_enum::E>(0)>>>
 // CHECK: template <> struct KernelInfo<::T1<::T3<::type_argument_template_enum::E>>>
 // CHECK: template <> struct KernelInfo<::dummy_functor_8<::EnumTypeOut, Baz>>
+
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/kernelname-enum.cpp:211]: Please revert this change.

File: clang/test/SemaSYCL/kernel-function-type.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -ast-dump %s | FileCheck %s
+// expected-no-diagnostics
+
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/kernel-function-type.cpp:3]: I asked about test description in test comments.

File: clang/test/SemaSYCL/kernel-function-type.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -ast-dump %s | FileCheck %s
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/kernel-function-type.cpp:1]: I would expect `-verify` flag.

Comment (erichkeane) [reply clang/test/SemaSYCL/kernel-function-type.cpp:1]: I agree, this shouldn't check the AST, it should just check the static-assert.

**Comment by erichkeane** [comment]
> 
> 
> First, I have seen a PR with similar changes - #1921, why you created a new one?
> Second, I was under impression that your solution tried to get rid of `getFullyQualifiedType` calls, in this patch you just left one on line 2194, why?

I think she is having git troubles (she was running off a private branch rather than a Fork) and decided to start a new one.

That said, the line 2194 use needs to go away.  I realize that changes some of the integration headers, but this patch ALSO needs to re-implement the original purpose (FQNs in integration headers) by altering the printing policy.

**Comment by zahiraam** [comment]
> First, I have seen a PR with similar changes - #1921, why you created a new one?
> Second, I was under impression that your solution tried to get rid of `getFullyQualifiedType` calls, in this patch you just left one on line 2194, why?

As I mentioned in our chat,  I was advised to use forks instead to push changes and furthermore my branch got polluted and I was unable to get it to a stable state.
The reason I didn't remove that getFullyQualifiedType at line 2194, was because it was breaking the LIT test in regression/kernel_name_inside_sycl_namespace.cpp. 
I am still trying to understand the reason.

File: clang/test/SemaSYCL/kernel-function-type.cpp (hunk 3 above)
Comment (zahiraam) [reply clang/test/SemaSYCL/kernel-function-type.cpp:1]: Should I remove the -ast-dump flag?

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -2164,7 +2160,11 @@ static void printArgument(ASTContext &Ctx, raw_ostream &ArgOS,
     TypePolicy.SuppressTypedefs = true;
     TypePolicy.SuppressTagKeyword = true;
     QualType T = Arg.getAsType();
-    ArgOS << getKernelNameTypeString(T, Ctx, TypePolicy);
+    if (T->isEnumeralType())
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2163]: Hmm... i would expect getKernelNameTypeString to just return the right value for the enum instead of needing ot handle it here.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 5]
```diff
@@ -2234,8 +2232,29 @@ static std::string getKernelNameTypeString(QualType T, ASTContext &Ctx,
 
     return eraseAnonNamespace(ArgOS.str().str());
   }
-
-  return eraseAnonNamespace(FullyQualifiedType.getAsString(TypePolicy));
+  if (const auto *D = dyn_cast<CXXRecordDecl>(RD)) {
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2235]: Does this work if it is a type defined in a lambda?  Something like:

    namespace NS {
      struct Foo {
          void bar() {
              []() {
                   struct Here{void operator()();};
                    }
        }
      };
    }

The type 'Here' is named something like: NS::Foo::bar()::{lambda()#1}::operator()() const::Here.  

Though, actually, I'm not sure how to forward declare that type... this might just be a general bug (kernels with names in an anonymous type).  @mkinsner @rolandschulz  what does the standard say about that?

File: clang/lib/Sema/SemaSYCL.cpp [hunk 6]
```diff
@@ -2234,8 +2232,29 @@ static std::string getKernelNameTypeString(QualType T, ASTContext &Ctx,
...
+    std::string NSStr = "";
+    const DeclContext *DC = D->getDeclContext();
+    SmallString<64> Buf;
+    llvm::raw_svector_ostream OS(Buf);
+    while (DC) {
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2240]: Instead of this, check out: https://clang.llvm.org/doxygen/classclang_1_1DeclContext.html#a52ebc03bfefa68fcda366dd5a7164c65

That way you can just do this in order.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 7]
```diff
@@ -2234,8 +2232,29 @@ static std::string getKernelNameTypeString(QualType T, ASTContext &Ctx,
...
+      auto *NS = dyn_cast<NamespaceDecl>(DC);
+      if (!NS || (NS && NS->getName().empty())) {
+        if (DC->isTranslationUnit())
+          NSStr.append(Twine("::" + D->getName()).str());
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2244]: The Twine function-like-cast is really odd here.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 8]
```diff
@@ -2234,8 +2232,29 @@ static std::string getKernelNameTypeString(QualType T, ASTContext &Ctx,
...
-
-  return eraseAnonNamespace(FullyQualifiedType.getAsString(TypePolicy));
+  if (const auto *D = dyn_cast<CXXRecordDecl>(RD)) {
+    std::string NSStr = "";
+    const DeclContext *DC = D->getDeclContext();
+    SmallString<64> Buf;
+    llvm::raw_svector_ostream OS(Buf);
+    while (DC) {
+      auto *NS = dyn_cast<NamespaceDecl>(DC);
+      if (!NS || (NS && NS->getName().empty())) {
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2242]: I don't really get what is happening here...

File: clang/lib/Sema/SemaSYCL.cpp (hunk 4 above)
Comment (zahiraam) [reply clang/lib/Sema/SemaSYCL.cpp:2163]: This happens where there is a call to getKernelNameTypeString from within the printTemplateArguments. Without this the kernelname-enum.cpp in dummy_functor_7<no_namespace_int> was not generating the "::". I initially put this in getKernelNameTypeString when RD is null, but that broke some other test cases where the "::" were added when they shouldn't. Let me look more into it.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 5 above)
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:2235]: If I am understanding this correctly, you are replacing the FullyQualifiedType because it causes type mismatch, but you still need fully qualified canonical name. Did I get that right? Does it make sense to move the calculation of fully qualified canonical name outside SemaSycl?

File: clang/lib/Sema/SemaSYCL.cpp (hunk 7 above)
Comment (zahiraam) [reply clang/lib/Sema/SemaSYCL.cpp]: Proposed a new solution to this.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 6 above)
Comment (zahiraam) [reply clang/lib/Sema/SemaSYCL.cpp:2240]: Not sure. I am assuming that you want me replace the while loop with something like this:
mallVector<DeclContext *, 8> Contexts;
    const_cast<DeclContext *>(D->getDeclContext())->collectAllContexts(Contexts);
    for (auto *DC : Contexts) {
// put content of while loop here
}
If that's what you thinking of then not sure that does what we want. For example for this:
namespace nm1 {
  namespace nm2 {
    class C {};
    class KernelName0 : public C {};
  } // namespace nm2
It will create a Contexts vector of size 1 containing only nm2. So that will not work.
Unless you are thinking of something else?

File: clang/test/SemaSYCL/kernel-function-type.cpp (hunk 3 above)
Comment (Fznamznon) [reply clang/test/SemaSYCL/kernel-function-type.cpp]: I think if you shouldn't check the AST, then you don't need the `-ast-dump` flag.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 5 above)
Comment (zahiraam) [reply clang/lib/Sema/SemaSYCL.cpp:2235]: Yes. That's  right. Not sure about putting outside of SemaSycl.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 4 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2163]: Still unresolved.  That said, I DO think that getKernelNameTypeString should just 'do the right thing' here, rather than having this special case.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 5 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2235]: I don't think it makes sense to move it outside of SemaSYCL.  The way this is printed is entirely the responsiblity of the integration header logic.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 6 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2240]: Ah, it appears that doesn't do what I thought it would do.  

I was suggesting it because the entire business here with NSStr seems incorrect, we should just be outputting directly to a stream rather than using a string.  It likely requires making a stack of items, then for-eaching over them instead.  String prepends are just unbelievably expensive.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 5 above)
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:2235]: The reason I suggested it is because it looks like the required functionality here is - 'Print a canonical name with it's namespace qualification'. I didn't think that necessarily had to be restricted as a integration header functionality. Thought it might be useful in other workflows as well. That said, if it doesn't make sense to move it out, I think you can move this logic out to a separate function accepting type and calculate the required qualified name. getKernelNameTypeString() can then just call this where ever the name is required. This will also resolve the issue being discussed in L2163

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2235]: I see what you mean.  I'm actually surprised the pretty-printer doesn't do that already now that you mention it.  I doubt we should be iterating through the decl-contexts, when that likely already happens elsewhere.

Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:2235]: Yea me too. Makes me wonder if there is some less obvious way to get this functionality from Clang API.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 4 above)
Comment (zahiraam) [reply clang/lib/Sema/SemaSYCL.cpp:2163]: This will have to be special cased if not here then in getKernelNameTypeString. It would be the case where RD is not a record and T is an Enum. Will post the code to see if you agree with it.

**Comment by Fznamznon** [comment]
What is the status of this?
Is it ready for new review?

**Comment by zahiraam** [comment]
> What is the status of this?
> Is it ready for new review?

Yes.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 9]
```diff
@@ -2202,40 +2198,51 @@ static void printTemplateArguments(ASTContext &Ctx, raw_ostream &ArgOS,
...
+      auto *NS = dyn_cast<NamespaceDecl>(RD->getDeclContext());
+      if (NS->isAnonymousNamespace())
+        RD->printQualifiedName(OS, TypePolicy, false);
+      else
+        RD->printQualifiedName(OS, TypePolicy, true);
+    } else
+      RD->printQualifiedName(OS, TypePolicy, true);
+    return eraseAnonNamespace(OS.str().str());
+  }
+  return eraseAnonNamespace(T.getCanonicalType().getAsString(TypePolicy));
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:2232]: A few suggestions to this place:
- No need to cast `RD` to `CXXRecordDecl`, it seems to be already a `CXXRecordDecl`, right?
- We can check that declaration context is a namespace using `dyn_cast`
- Just a bit less of ifs
```suggestion
  if (RD->getDeclContext()->isFunctionOrMethod())
    return eraseAnonNamespace(T.getCanonicalType().getAsString(TypePolicy));
  if (auto *NS = cast<NamespaceDecl>(RD->getDeclContext()))
    RD->printQualifiedName(OS, TypePolicy, !NS->isAnonymousNamespace());
  else
    RD->printQualifiedName(OS, TypePolicy, /*WithGlobalNsPrefix=*/true);
  return eraseAnonNamespace(OS.str().str());
```

File: clang/test/SemaSYCL/kernel-function-type.cpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -verify -ast-dump %s | FileCheck %s
```
[the remark this answers is not in the document]
Comment (zahiraam) [reply clang/test/SemaSYCL/kernel-function-type.cpp:1]: The AST checking was to make sure that the right type is generated. But if you think it is redundant I can remove it.

[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/test/SemaSYCL/kernel-function-type.cpp:1]: I guess it is not needed, see https://github.com/intel/llvm/pull/1949#discussion_r444189442

File: clang/lib/Sema/SemaSYCL.cpp (hunk 9 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: This one still is not resolved.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 11]
```diff
@@ -2202,40 +2198,45 @@ static void printTemplateArguments(ASTContext &Ctx, raw_ostream &ArgOS,
...
     const TemplateArgumentList &Args = TSD->getTemplateArgs();
-    printTemplateArguments(Ctx, ArgOS, Args.asArray(), TypePolicy);
+    printTemplateArguments(Ctx, OS, Args.asArray(), TypePolicy);
+
+    return eraseAnonNamespace(OS.str().str());
+  }
+  if (RD->getDeclContext()->isFunctionOrMethod())
+    return eraseAnonNamespace(T.getCanonicalType().getAsString(TypePolicy));
+  if (dyn_cast<NamespaceDecl>(RD->getDeclContext()) &&
+      dyn_cast<NamespaceDecl>(RD->getDeclContext())->isAnonymousNamespace())
```
Comment (premanandrao) [inline clang/lib/Sema/SemaSYCL.cpp:2222]: You could avoid the second dynamic cast by storing the result of the first one.

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2222]: Yeah, you should also only be dyn_casting something if you're going to use the result, otherwise use isa.

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2222]: Also, odd to do this with an 'if/else' rather than just putting the condition in the call.

Comment (zahiraam) [reply clang/lib/Sema/SemaSYCL.cpp]: With the new version I have I think responded to Mariya and Prem's review. Not sure what you mean Erich? Do you mean the caller of printRecordType should be called with the Namespace? Or do you mean the call of printQualified?

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp]: I mean, it seems that the body of these two differs only by 'true' or 'false'.  Instead of the if/else, do:

    RD->printQualifiedName(OS, TypePolicy, NS && NS->isAnonymousNamespace());

File: clang/lib/Sema/SemaSYCL.cpp [hunk 12]
```diff
@@ -2202,40 +2198,42 @@ static void printTemplateArguments(ASTContext &Ctx, raw_ostream &ArgOS,
...
+    TSD->printQualifiedName(OS, TypePolicy, /*WithGlobalNsPrefix*/ true);
 
     // Print template arguments substituting enumerators
     ASTContext &Ctx = RD->getASTContext();
     const TemplateArgumentList &Args = TSD->getTemplateArgs();
-    printTemplateArguments(Ctx, ArgOS, Args.asArray(), TypePolicy);
+    printTemplateArguments(Ctx, OS, Args.asArray(), TypePolicy);
 
-    return eraseAnonNamespace(ArgOS.str().str());
+    return eraseAnonNamespace(OS.str().str());
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2217]: not your fault (and perhaps for our viewers), this is the reason why I hate the interface to eraseAnonNamespace.  It already makes a copy of the string, yet still takes a std::string by copy.  While it is acceptable int he case of xvalues (like this), it creates this awful interface.

IMO, it should take a stringref and build build a new one using Twines.

Comment (zahiraam) [reply clang/lib/Sema/SemaSYCL.cpp:2217]: I agree. Should it be re-written?

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2217]: Not in this patch.  Feel free to copy me a on PR to rewrite it.

**Comment by zahiraam** [comment]
I can't reproduce the error in the buildbot.
Here are the commands I ran:

zahiraam@scsel-cfl-02:/export/users/zahiraam/Testing$ /export/users/zahiraam/intel_fork/prodbuild/bin/clang --driver-mode=g++ -fsycl -fsycl-targets=spir64-unknown-linux-sycldevice /export/users/zahiraam/intel_fork/llvm/sycl/test/basic_tests/handler/handler_mem_op.cpp -o /export/users/zahiraam/intel_fork/prodbuild/tools/sycl/test/basic_tests/handler/Output/handler_mem_op.cpp.tmp.out

zahiraam@scsel-cfl-02:/export/users/zahiraam/Testing$ env SYCL_DEVICE_TYPE=HOST /export/users/zahiraam/intel_fork/prodbuild/tools/sycl/test/basic_tests/handler/Output/handler_mem_op.cpp.tmp.out
finish

zahiraam@scsel-cfl-02:/export/users/zahiraam/Testing$ true SYCL_DEVICE_TYPE=HOST /export/users/zahiraam/intel_fork/prodbuild/tools/sycl/test/basic_tests/handler/Output/handler_mem_op.cpp.tmp.out

zahiraam@scsel-cfl-02:/export/users/zahiraam/Testing$ env SYCL_DEVICE_TYPE=GPU SYCL_BE=PI_LEVEL0 /export/users/zahiraam/intel_fork/prodbuild/tools/sycl/test/basic_tests/handler/Output/handler_mem_op.cpp.tmp.out
finish
zahiraam@scsel-cfl-02:/export/users/zahiraam/Testing$

Any suggestions to find out more? Thanks.
